implode.go

```diff
@@ -0,0 +1,134 @@
+// ensuring ordered map functionality. When imploding map types, please utilize the OrderedMap type from the [orderedmap library]
+// to maintain element order. If you use map type, not OrderedMap type, the order of the results cannot be guaranteed.
+//
+// reference:
+//   - implode: https://github.com/php/php-src/blob/php-5.6.40/ext/standard/string.c#L1229-L1269
+//   - php_implode: https://github.com/php/php-src/blob/php-5.6.40/ext/standard/string.c#L1141-L1224
+//
+// Test cases:
+//   - https://github.com/php/php-src/blob/php-5.6.40/ext/standard/tests/strings/implode.phpt
+//   - https://github.com/php/php-src/blob/php-5.6.40/ext/standard/tests/strings/implode1.phpt
+//
+// [official PHP documentation]: https://www.php.net/manual/en/function.implode.php
+// [orderedmap library]: https://pkg.go.dev/github.com/elliotchance/orderedmap/v2
+func Implode(arg1 any, options ...any) (string, error) {
+	var delim string
+	var arr []any
+
+	// Check arg1 is one of array, slice, map, or ordered ap
+	isArg1CollectionType := isCollectionType(arg1)
+
+	// Check if options is not provided
+	if len(options) == 0 {
+		if !isArg1CollectionType {
+			return "", fmt.Errorf("argument must be one of array, slice, or ordered map, but got %v", reflect.TypeOf(arg1))
+		}
+		aggregateValues(arg1, &arr)
+	} else {
+		arg2 := options[0]
+		// Check arg2 is one of array, slice, or ordered map
+		isArg2CollectionType := isCollectionType(arg2)
+
+		if isArg1CollectionType {
+			delim, _ = ConvertToString(arg2)
+			aggregateValues(arg1, &arr)
+		} else if !isArg1CollectionType && isArg2CollectionType {
+			delim, _ = ConvertToString(arg1)
+			aggregateValues(arg2, &arr)
+		} else {
+			return "", fmt.Errorf("invalid arguments passed, got %v, %v", reflect.TypeOf(arg1), reflect.TypeOf(arg2))
+		}
+	}
+
+	// Join arr elements with a delim
+	var builder strings.Builder
+	if len(arr) == 0 {
+		return "", nil
+	}
+	for i, item := range arr {
+		str, err := ConvertToString(item)
+
+		if err != nil {
+			return "", fmt.Errorf("unsupported type in array : %v", reflect.TypeOf(item))
+		} else {
+			builder.WriteString(str)
+		}
+		if i < len(arr)-1 {
+			builder.WriteString(delim)
+		}
+	}
+	return builder.String(), nil
+}
+
+// isOrderedMap checks if the argument is an instance of ordered map
+func isOrderedMap(arg any) bool {
+	_, ok := arg.(*orderedmap.OrderedMap[any, any])
+	return ok
+}
+
+// isCollectionType checks if the argument is either an array, a slice, a map or ordered map
+func isCollectionType(arg any) bool {
+	if arg == nil {
+		return false
+	}
+
+	argType := reflect.TypeOf(arg).Kind()
+	return isOrderedMap(arg) || argType == reflect.Slice || argType == reflect.Array || argType == reflect.Map
+}
+
+// aggregateValues extracts the stored value from different types of source:
+// ordered map, map, slice and array. It gathers there values into an arr and returns it.
+func aggregateValues(source any, arr *[]any) {
```

Comment:
As far as I know, the slice is reference type. So even though slice is value, internally slice hold pointer to the memory space which store data. 

And, it looks the `arr` is not modified many times. So I think it would be better if this function initialized and returned `arr`. 

- [Pointer and slice reference type - receiver](https://stackoverflow.com/questions/29838240/pointer-and-slice-reference-type-receiver)

Reply:
I'v refactored the aggregateValues function to return the slice that collected the source values.